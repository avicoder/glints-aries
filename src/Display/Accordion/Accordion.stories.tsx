import React from 'react';
import { Story, Meta } from '@storybook/react';

import Accordion, { Props as AccordionProps } from './Accordion';
import { Props as AccordionPanelProps } from './AccordionPanel';
import { BaseContainer } from '../../Layout/GlintsContainer/GlintsContainer';

export default {
  title: 'Display/Accordion',
  component: Accordion,
  argTypes: {
    className: {
      table: {
        disable: true,
      },
    },
    label: {
      description: 'Sets the label for <Accordion.Panel />',
      control: {
        type: 'text',
      },
    },
    content: {
      description: 'Sets the content for <Accordion.Panel />',
      control: {
        type: 'text',
      },
    },
  },
  decorators: [Story => <BaseContainer>{Story()}</BaseContainer>],
} as Meta;

const Template: Story<AccordionProps & AccordionPanelProps> = ({
  iconPosition,
  label = 'What is Lorem Ipsum?',
  content = 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.',
}) => (
  <Accordion iconPosition={iconPosition}>
    <Accordion.Panel label={label} content={content} />
  </Accordion>
);

export const Interactive = Template.bind({});
Interactive.args = {
  iconPosition: 'left',
};

export const DifferentIconPosition = Template.bind({});
DifferentIconPosition.args = {
  iconPosition: 'right',
};
