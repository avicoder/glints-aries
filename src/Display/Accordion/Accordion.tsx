import * as React from 'react';
import classNames from 'classnames';

import {
  AddIcon,
  MinusIcon,
  ArrowDownSolidIcon,
} from '../../General/Icon/components';
import AccordionPanel, { Props as AccordionPanelProps } from './AccordionPanel';

import { Container } from './AccordionStyle';

const Accordion: Accordion = ({
  children,
  className,
  iconPosition = 'left',
}) => {
  const [currIndex, setCurrIndex] = React.useState(-1);
  const iconOptions: IconOptions = {
    activeIcon:
      iconPosition === 'left' ? <MinusIcon /> : <ArrowDownSolidIcon />,
    inactiveIcon:
      iconPosition === 'left' ? <AddIcon /> : <ArrowDownSolidIcon />,
    position: iconPosition,
  };

  const handleOpen = (index: number) => {
    setCurrIndex(currIndex === index ? -1 : index);
  };

  return (
    <Container className={classNames('aries-accordion', className)}>
      {React.Children.map(
        children,
        (child: React.ReactElement<AccordionPanelProps>, index) => {
          const { label, content, ...restChildProps } = child.props;
          return React.cloneElement(child, {
            key: index,
            label: label,
            content: content,
            active: currIndex === index,
            iconOptions: iconOptions,
            onOpen: () => handleOpen(index),
            ...restChildProps,
          });
        }
      )}
    </Container>
  );
};

type Accordion = React.FunctionComponent<Props> & {
  Panel: typeof AccordionPanel;
};

export type IconPosition = 'left' | 'right';

export interface IconOptions {
  activeIcon: React.ReactElement;
  inactiveIcon: React.ReactElement;
  position: IconPosition;
}

export interface Props {
  children: React.ReactNode;
  className?: string;
  /** Sets the position of the icon. */
  iconPosition?: IconPosition;
}

Accordion.Panel = AccordionPanel;
Accordion.Panel.displayName = 'Accordion.Panel'; // To show the correct name in code sample

export default Accordion;
