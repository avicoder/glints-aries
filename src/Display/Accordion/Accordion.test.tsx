import '@testing-library/jest-dom/extend-expect';
import * as React from 'react';
import { render } from '@testing-library/react';

import Accordion, { IconPosition } from './Accordion';
import userEvent from '@testing-library/user-event';

describe('<Accordion /> prop className', () => {
  const matchSnapshot = (className: string) => {
    test(`should match snapshot when class name ${className} is passed`, () => {
      const { asFragment } = render(
        <Accordion className={className}>
          <Accordion.Panel
            className={className}
            label="label"
            content="content"
          />
        </Accordion>
      );
      // validate class name from snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  };

  ['test', undefined].forEach(matchSnapshot);
});

describe('<Accordion.Panel /> prop iconPosition', () => {
  const iconPositionArray = ['left', 'right'] as IconPosition[];
  const matchSnapshot = (iconPosition: IconPosition) => {
    test(`should match snapshot when iconPosition ${iconPosition} is passed`, () => {
      const { asFragment } = render(
        <Accordion iconPosition={iconPosition}>
          <Accordion.Panel label="label" content="content" />
        </Accordion>
      );
      // validate icon position from snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  };

  iconPositionArray.forEach(iconPosition => matchSnapshot(iconPosition));
});

describe('<Accordion /> click behavior', () => {
  test('<Accordion.Panel /> should be open/ close when it is clicked', () => {
    const { queryByText } = render(
      <Accordion>
        <Accordion.Panel label="label" content="content" />
      </Accordion>
    );

    const panelLabel = queryByText('label');
    const panelContent = queryByText('content');
    expect(panelContent).not.toBeVisible();

    userEvent.click(panelLabel);
    expect(panelContent).toBeVisible();

    userEvent.click(panelLabel);
    expect(panelContent).not.toBeVisible();
  });

  test('<Accordion /> should open one panel only', () => {
    const { queryByText } = render(
      <Accordion>
        <Accordion.Panel label="label-1" content="content-1" />
        <Accordion.Panel label="label-2" content="content-2" />
      </Accordion>
    );

    const panelContent1 = queryByText('content-1');
    const panelContent2 = queryByText('content-2');
    expect(panelContent1).not.toBeVisible();
    expect(panelContent2).not.toBeVisible();

    const panelLabel1 = queryByText('label-1');
    userEvent.click(panelLabel1);

    expect(panelContent1).toBeVisible();
    expect(panelContent2).not.toBeVisible();

    const panelLabel2 = queryByText('label-2');
    userEvent.click(panelLabel2);

    expect(panelContent1).not.toBeVisible();
    expect(panelContent2).toBeVisible();
  });
});
