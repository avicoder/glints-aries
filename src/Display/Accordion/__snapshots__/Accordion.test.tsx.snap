// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> prop className should match snapshot when class name test is passed 1`] = `
<DocumentFragment>
  <div
    class="AccordionStyle__Container-sc-7hk6zs-1 cmGSOk aries-accordion test"
  >
    <div
      aria-expanded="false"
      class="AccordionStyle__PanelWrapper-sc-7hk6zs-0 cGRhuq panel-wrapper test"
      role="tab"
      tabindex="0"
    >
      <div
        class="AccordionStyle__IconLabelWrapper-sc-7hk6zs-2 NwhRw label-wrapper"
        tabindex="-1"
      >
        <div
          class="AccordionStyle__IconWrapper-sc-7hk6zs-6 fHDKbm"
        >
          <svg
            class="IconStyle__VerticalCenteredSvg-x7d280-0 iFpcQN"
            data-testid="icon-svg"
            fill="currentColor"
            height="1em"
            viewBox="0 0 100 100"
            width="1em"
          >
            <path
              d="M100 56.563H56.562V100H43.438V56.562H0V43.438h43.438V0h13.124v43.438H100z"
            />
          </svg>
        </div>
        <div
          class="AccordionStyle__Label-sc-7hk6zs-5 loNlGz"
        >
          label
        </div>
      </div>
      <div
        class="AccordionStyle__ContentWrapper-sc-7hk6zs-3 fbmYVN content-wrapper"
      >
        <div
          class="AccordionStyle__Content-sc-7hk6zs-4 hPMhus"
        >
          content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Accordion /> prop className should match snapshot when class name undefined is passed 1`] = `
<DocumentFragment>
  <div
    class="AccordionStyle__Container-sc-7hk6zs-1 cmGSOk aries-accordion"
  >
    <div
      aria-expanded="false"
      class="AccordionStyle__PanelWrapper-sc-7hk6zs-0 cGRhuq panel-wrapper"
      role="tab"
      tabindex="0"
    >
      <div
        class="AccordionStyle__IconLabelWrapper-sc-7hk6zs-2 NwhRw label-wrapper"
        tabindex="-1"
      >
        <div
          class="AccordionStyle__IconWrapper-sc-7hk6zs-6 fHDKbm"
        >
          <svg
            class="IconStyle__VerticalCenteredSvg-x7d280-0 iFpcQN"
            data-testid="icon-svg"
            fill="currentColor"
            height="1em"
            viewBox="0 0 100 100"
            width="1em"
          >
            <path
              d="M100 56.563H56.562V100H43.438V56.562H0V43.438h43.438V0h13.124v43.438H100z"
            />
          </svg>
        </div>
        <div
          class="AccordionStyle__Label-sc-7hk6zs-5 loNlGz"
        >
          label
        </div>
      </div>
      <div
        class="AccordionStyle__ContentWrapper-sc-7hk6zs-3 fbmYVN content-wrapper"
      >
        <div
          class="AccordionStyle__Content-sc-7hk6zs-4 hPMhus"
        >
          content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Accordion.Panel /> prop iconPosition should match snapshot when iconPosition left is passed 1`] = `
<DocumentFragment>
  <div
    class="AccordionStyle__Container-sc-7hk6zs-1 cmGSOk aries-accordion"
  >
    <div
      aria-expanded="false"
      class="AccordionStyle__PanelWrapper-sc-7hk6zs-0 cGRhuq panel-wrapper"
      role="tab"
      tabindex="0"
    >
      <div
        class="AccordionStyle__IconLabelWrapper-sc-7hk6zs-2 NwhRw label-wrapper"
        tabindex="-1"
      >
        <div
          class="AccordionStyle__IconWrapper-sc-7hk6zs-6 fHDKbm"
        >
          <svg
            class="IconStyle__VerticalCenteredSvg-x7d280-0 iFpcQN"
            data-testid="icon-svg"
            fill="currentColor"
            height="1em"
            viewBox="0 0 100 100"
            width="1em"
          >
            <path
              d="M100 56.563H56.562V100H43.438V56.562H0V43.438h43.438V0h13.124v43.438H100z"
            />
          </svg>
        </div>
        <div
          class="AccordionStyle__Label-sc-7hk6zs-5 loNlGz"
        >
          label
        </div>
      </div>
      <div
        class="AccordionStyle__ContentWrapper-sc-7hk6zs-3 fbmYVN content-wrapper"
      >
        <div
          class="AccordionStyle__Content-sc-7hk6zs-4 hPMhus"
        >
          content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Accordion.Panel /> prop iconPosition should match snapshot when iconPosition right is passed 1`] = `
<DocumentFragment>
  <div
    class="AccordionStyle__Container-sc-7hk6zs-1 cmGSOk aries-accordion"
  >
    <div
      aria-expanded="false"
      class="AccordionStyle__PanelWrapper-sc-7hk6zs-0 cGRhuq panel-wrapper"
      role="tab"
      tabindex="0"
    >
      <div
        class="AccordionStyle__IconLabelWrapper-sc-7hk6zs-2 gDlQm label-wrapper"
        tabindex="-1"
      >
        <div
          class="AccordionStyle__Label-sc-7hk6zs-5 loNlGz"
        >
          label
        </div>
        <div
          class="AccordionStyle__IconWrapper-sc-7hk6zs-6 fmxLTd"
        >
          <svg
            class="IconStyle__VerticalCenteredSvg-x7d280-0 iFpcQN"
            data-testid="icon-svg"
            fill="currentColor"
            height="1em"
            viewBox="0 0 100 100"
            width="1em"
          >
            <path
              d="M100 25L50 75 0 25z"
            />
          </svg>
        </div>
      </div>
      <div
        class="AccordionStyle__ContentWrapper-sc-7hk6zs-3 fbmYVN content-wrapper"
      >
        <div
          class="AccordionStyle__Content-sc-7hk6zs-4 eEHaoA"
        >
          content
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
